### PR TITLE
Close #117 add a ADLB template

### LINKED ISSUE #117: Template: ADLB
### Feature Idea

Feature Idea
Background Information
Showcase from start to finish how to create an ADLB. This should include:

Parameter mappings​
Endpoint programming​
Should be based on LB

Definition of Done

New file inst/templates/adlb.R showcasing how to create ADLB

Relevant Input
No response

Relevant Output
No response

Reproducible Example/Pseudo Code
No response

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by meilinjiang**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral family codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral<ext> (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Review by jwang-lilly — APPROVED** [review]

**Comment by jwang-lilly** [comment]
@meilinjiang, all look great! You may need to rerun the styler, commit and push again. Oh, check off all the boxes.

File: inst/templates/ad_adlb.R [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+lb_neuro <- pharmaversesdtm::lb_neuro
+admiralneuro_adsl <- admiralneuro::adsl_neuro
+
+# When SAS datasets are imported into R using haven::read_sas(), missing
+# character values from SAS appear as "" characters in R, instead of appearing
+# as NA values. Further details can be obtained via the following link:
+# https://pharmaverse.github.io/admiral/articles/admiral.html#handling-of-missing-values # nolint
+adlb <- convert_blanks_to_na(lb_neuro)
+adsl <- convert_blanks_to_na(admiralneuro_adsl)
+
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R:31]: Please delete extra spaces

File: inst/templates/ad_adlb.R [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+          is.na(AVAL) | as.character(signif(LBSTRESN2,5)) != LBSTRESC,
+          LBSTRESC,
+          NA_character_
+        ),
+        ANRLO = LBSTNRLO,
+        ANRHI = LBSTNRHI
+)
+
+## Domain-specific derivations ----
+
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R:122]: Please delete extra spaces

File: inst/templates/ad_adlb.R [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  derive_var_obs_number(
+    new_var = ASEQ,
+    by_vars = get_admiral_option("subject_keys"),
+    order = exprs(PARAMCD, ADT, AVISITN, VISITNUM),
+    check_type = "error"
+  )
+
+# Final Steps, Select final variables and Add labels ----
+# This process will be based on your metadata, no example given for this reason
+
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R:223]: Please delete extra spaces

File: inst/templates/ad_adlb.R [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+  ## Join ADSL with LB (need TRTSDT for ADY derivation) ----
+derive_vars_merged(
+  dataset_add = adsl,
+  new_vars = adsl_vars,
+  by_vars = get_admiral_option("subject_keys")
+)
+
+adlb <- adlb %>%
+  ## Add PARAMCD and PARAM ----
+derive_vars_merged_lookup(
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]: Please run `styler()`

File: inst/templates/ad_adlb.R [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+# Use the admiral option functionality to store subject key variables in one
+# place (note: `subject_keys` defaults to STUDYID and USUBJID)
+set_admiral_options(subject_keys = exprs(STUDYID, USUBJID))
+
+# Load source datasets ----
+
+# Use e.g. `haven::read_sas()` to read in .sas7bdat, or other suitable functions
+# as needed and assign to the variables below.
+# For illustration purposes read in pharmaversesdtm and admiralneuro neuro test data
+# nv <- pharmaversesdtm::nv_neuro
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]: Please delete unnecessary code

**Review by Fanny-Gautier — CHANGES_REQUESTED** [review]
Please implement below comments, run `styler()` and update `NEWS.md` accordingly.

**Comment by meilinjiang** [comment]
> @meilinjiang, all look great! You may need to rerun the styler, commit and push again. Oh, check off all the boxes.

Addressed

File: inst/templates/ad_adlb.R (hunk 1 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R:31]: Done

File: inst/templates/ad_adlb.R (hunk 2 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R:122]: Done

File: inst/templates/ad_adlb.R (hunk 3 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R:223]: Done

File: inst/templates/ad_adlb.R (hunk 4 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: Done

File: inst/templates/ad_adlb.R (hunk 5 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: Unnecessary code and comments removed

File: inst/templates/ad_adlb.R [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    new_vars = exprs(PARAMCD, PARAM, PARAMN),
+    by_vars = exprs(LBTESTCD)
+  )
+
+# Derive Date/Time and Analysis Day ----
+# See the "Derive/Impute Numeric Date/Time and Analysis Day" vignette section
+# for more information:
+# (https://pharmaverse.github.io/admiral/articles/bds_finding.html#datetime)
+
+# Add analysis date (ADT) and treatment start date (TRTSDT)
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]: Please update comment: Treatment start date was added line 50, not here.

File: inst/templates/ad_adlb.R [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+      is.na(AVAL) | as.character(signif(LBSTRESN2, 5)) != LBSTRESC,
+      LBSTRESC,
+      NA_character_
+    ),
+    ANRLO = LBSTNRLO,
+    ANRHI = LBSTNRHI
+  )
+
+## Domain-specific derivations ----
+# To be added in the future for more parameters
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]: Could you clarify the comment? Is it a note for yourself to add something later, or is it meant for the user? I think it could be dropped.

File: inst/templates/ad_adlb.R [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+  derive_var_ontrtfl(
+    start_date = ADT,
+    ref_start_date = TRTSDT,
+    ref_end_date = TRTEDT,
+    filter_pre_timepoint = toupper(AVISIT) == "BASELINE" # Observations as not on-treatment
+  )
+
+### Derive Baseline flags ----
+
+### Calculate ABLFL ----
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]: Please review the structure of the TOC/comments. Repetition of line 131?

File: inst/templates/ad_adlb.R [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    args = params(
+      new_var = ABLFL,
+      by_vars = c(get_admiral_option("subject_keys"), exprs(BASETYPE, PARAMCD)),
+      order = exprs(ADT, VISITNUM, LBSEQ),
+      mode = "last"
+    ),
+    filter = ((!is.na(AVAL) | !is.na(AVALC)) & ADT <= TRTSDT & !is.na(BASETYPE))
+  )
+
+## Derive visit flags ----
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]: Did you mean analysis flags that are based on the visit?

File: inst/templates/ad_adlb.R [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    filter = AVISITN > 0
+  ) %>%
+  ### Calculate PCHG for post-baseline records ----
+  # The decision on how to populate pre-baseline and baseline values of PCHG is left to producer choice
+  restrict_derivation(
+    derivation = derive_var_pchg,
+    filter = AVISITN > 0
+  )
+
+## Assign ASEQ ----
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]: Please add that it is an optional variable
```suggestion
## Assign ASEQ (Optional Variable) ----
```

File: inst/templates/ad_adlb.R [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  )
+
+# Derive results ----
+# See the "Derive Results (AVAL, AVALC)" vignette section for more information:
+# (https://pharmaverse.github.io/admiral/articles/bds_finding.html#aval)
+
+# Derive AVAL and AVALC and define parameter categories
+adlb <- adlb %>%
+  mutate(
+    LBSTRESN2 = case_when(
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R:99]: What is `LBSTRESN2` ? This is not CDISC compliant being kept in final dataset.

File: inst/templates/ad_adlb.R [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+      PARAMN == 1 ~ round(LBSTRESN, 4),
+      PARAMN == 2 ~ round(LBSTRESN, 1),
+      PARAMN == 3 ~ round(LBSTRESN, 5),
+      PARAMN == 4 ~ LBSTRESN,
+      TRUE ~ NA
+    ),
+    AVAL = LBSTRESN,
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R:106]: Format of `AVAL` is not correct, some values are displayed as `1.120311e+03`.
Please also check `LB.LBSTRESN` format for the same.

**Review by Fanny-Gautier — CHANGES_REQUESTED** [review]
A few more comments to implement. Thank you for this template!

File: NEWS.md [hunk 13]
```diff
@@ -23,6 +23,7 @@ function also now only accepts positive SUVR values. (#106)
 ## Template programs
 
 - ADNV template scripts `ad_adnv.R` which creates a Nervous System Analysis Dataset. (#113)
+- ADLB template scripts `ad_adlb.R` which creates a Lab Analysis Dataset. (#119)
```
Comment (Fanny-Gautier) [inline NEWS.md]: This should be added within the development version section

File: inst/templates/ad_adlb.R [hunk 14]
```diff
@@ -0,0 +1,221 @@
+# Name: ADLB
+#
+# Label: Laboratory Analysis dataset for neuroscience
```
Comment (Fanny-Gautier) [inline inst/templates/ad_adlb.R]:
```suggestion
# Label: Laboratory Analysis Dataset for Neuroscience
```

File: NEWS.md (hunk 13 above)
Comment (Fanny-Gautier) [inline NEWS.md]:
```suggestion
- ADLB template scripts `ad_adlb.R` which creates a Laboratory Analysis Dataset. (#119)
```

File: inst/templates/ad_adlb.R (hunk 11 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R:99]: LBSTRESN2 is not CDISC but used as an intermediate step to derive AVAL and AVALC given the special rounding format of LBSTRESN in our lb_neuro data. I have modified code to drop it after the derivation.

File: inst/templates/ad_adlb.R (hunk 12 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R:106]: added options(scipen = 999) #turn off scientific notation

File: NEWS.md (hunk 13 above)
Comment (meilinjiang) [reply NEWS.md]: Added a new section # admiralneuro 0.2.2.9000. Any thoughts on the new development version number?

File: inst/templates/ad_adlb.R (hunk 14 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: fixed

File: NEWS.md (hunk 13 above)
Comment (meilinjiang) [reply NEWS.md]: fixed

File: inst/templates/ad_adlb.R (hunk 10 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: fixed

File: inst/templates/ad_adlb.R (hunk 9 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: Yes and fixed

File: inst/templates/ad_adlb.R (hunk 8 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: Combined to one line

File: inst/templates/ad_adlb.R (hunk 7 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: It is a note for our developer team. As discussed, there is no currently domain-specific derivations in the current adlb template, but we can add in the future once more test data come in if needed. I have removed this section for now to avoid confusion

File: inst/templates/ad_adlb.R (hunk 6 above)
Comment (meilinjiang) [reply inst/templates/ad_adlb.R]: fixed

File: NEWS.md [hunk 15]
```diff
@@ -1,3 +1,9 @@
+# admiralneuro 0.2.2.9000
+
+## Template programs
+
+- ADLB template scripts `ad_adlb.R` which creates a Laboratory Analysis Dataset. (#119)
+
 # admiralneuro 0.2.1.9000
```
Comment (Fanny-Gautier) [inline NEWS.md]: I believe this should be `# admiralneuro 0.2.1`. Is it the released one?

File: NEWS.md [hunk 16]
```diff
@@ -1,3 +1,9 @@
+# admiralneuro 0.2.2.9000
```
Comment (Fanny-Gautier) [inline NEWS.md]: If this this the current development version it should be `# admiralneuro (development version)`
